### PR TITLE
sync permissions / assignable_by to couch directly

### PR DESCRIPTION
**Pull request by snopoke**
## Summary
To avoid issues where changes to roles are only partially synced call the sync directly when changing permissions / assignable_by.

See https://github.com/dimagi/commcare-hq/pull/29971#discussion_r660023926

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
Covered by existing tests.

### Safety story
This removes the need to call `role._migration_do_sync` after updating related models. The sync can be disabled by passing `sync_to_couch=False` but this is only done in selected places where it is safe to do so:
- when updating the SQL role from couch
- when creating new roles (the sync happens after the SQL role is fully created).

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/users/models_sql.py [hunk 1]
```diff
@@ -99,11 +99,11 @@ def create(cls, domain, name, permissions=None, assignable_by=None, **kwargs):
             if permissions is None:
                 # match couch functionality and set default permissions
                 permissions = Permissions()
-            role.set_permissions(permissions.to_list())
+            role.set_permissions(permissions.to_list(), sync_to_couch=False)
```
Comment (millerdev) [inline corehq/apps/users/models_sql.py:102]: It looks like `set_permissions` is a similar method to `set_assignable_by` where the same pattern is being implemented. Is that right? Did that have the same issue as `set_assignable_by`, and now it is being fixed?

nit: if all that is true it would have been nice to split that out in a separate commit. (no change needed now)

**Review by millerdev — COMMENTED** [review]
> Covered by existing tests.

It seems like this PR is adding new features: sync to couch in `set_permissions` and `set_assignable_by`. Assuming no tests were failing before, it doesn't seem like these features are covered by existing tests. What do you think of adding tests that verify the couch sync when each of those methods is called?

File: corehq/apps/users/models_sql.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/users/models_sql.py:102]: yes, that is true. Both methods create related models that were not synced to couch during the method call. I will remember to try split changes up in future.

**Comment by snopoke** [comment]
> It seems like this PR is adding new features: sync to couch in set_permissions and set_assignable_by. Assuming no tests were failing before, it doesn't seem like these features are covered by existing tests. What do you think of adding tests that verify the couch sync when each of those methods is called?

Added a specific test for each: d01d72a30ca, b83210f797d

File: corehq/apps/users/tests/test_migrate_roles_to_sql.py [hunk 2]
```diff
@@ -98,6 +97,49 @@ def test_sync_role_sql_to_couch(self):
...
+            domain=self.domain,
+            name="test_sql_to_couch",
+            default_landing_page=ALL_LANDING_PAGES[0].id,
+            is_non_admin_editable=False,
+            upstream_id=self.app_editor_sql.couch_id
+        )
+        sql_role.save()
+        couch_roles = UserRole.by_domain(self.domain)
+        self.assertEqual(len(couch_roles), 2)
+        couch_role = [r for r in couch_roles if r.name == sql_role.name][0]
```
Comment (millerdev) [inline corehq/apps/users/tests/test_migrate_roles_to_sql.py:111]: This is hard to understand without extra context. Why are there 2 couch roles? Is that part of the test setup? Why is the first one special? Would it be possible to change it identify roles by name rather than magic numbers? Something like
```suggestion
        couch_roles = {r.name: r for r in UserRole.by_domain(self.domain)}
        self.assertEqual(couch_roles.keys(), {"test_sql_to_couch", "something_else"})  # <-- the second one is certainly incorrect
        couch_role = couch_roles["test_sql_to_couch"]
```

File: corehq/apps/users/tests/test_migrate_roles_to_sql.py [hunk 3]
```diff
@@ -98,6 +97,49 @@ def test_sync_role_sql_to_couch(self):
...
```
Comment (millerdev) [inline corehq/apps/users/tests/test_migrate_roles_to_sql.py:135]: Same as above comment.

**Review by millerdev — APPROVED** [review]

**Comment by millerdev** [comment]
Thanks for adding tests.

File: corehq/apps/users/tests/test_migrate_roles_to_sql.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/users/tests/test_migrate_roles_to_sql.py]: > This is hard to understand without extra context. Why are there 2 couch roles? Is that part of the test setup? Why is the first one special? Would it be possible to change it identify roles by name rather than magic numbers? Something like

There is one role created during setup. I've updated the tests to fetch the role under test directly instead of via the view: ed09c94e078

**Review by millerdev — APPROVED** [review]
